### ISSUE TITLE
When I manipulated the device's time into the past, Offline Playback continues despite LicenseDuration expiration.

### ISSUE THREAD

**Issue by brosvaby** [google/ExoPlayer#2622]
### Issue description
the bottom line is that when I manipulated the device time into the past, mediadrm have not detected a change in device time. Playback continues despite expiration.

I succeeded in offline license acquisition and content playback using widevine drm.
We set the license period to 300 sec and confirmed LicenseDurationRemaining to 300 sec. In normal cases, a period expiration event occurs after 300 seconds and playback does not occur. However, during license has not been deleted, we have confirmed that if the device time is changed to 5 minutes ago, LicenseDurationRemaining increases and the playback continues even though the time to expire has passed.

The bottom line is that we have not detected a change in device time.

### Reproduction steps
1. download offline license.
2. play the content offline
3. Normally, confirm that LicenseDurationRemaining has been getting reduced.
4. at your device : Settings -> Date & time -> turn off Auto-date & time
5. at your device : Settings -> Date & time -> Set date -> turn back time
6. Make sure that playback continues after the expiration time.

### Version of ExoPlayer being used
2.3.0

### Device(s) and version(s) of Android being used
device : All devices 
version : 5.0 and over (5.0, 6.0, 7.0)

**Comment by brosvaby** [comment]
Is anything updated?
For the above reasons, I think that we can't control period of expiration offline.
Do I misunderstand offline scenarios?
For example, is it necessary to control in the app area for device time manipulation?

**Comment by wvpaf** [comment]
Hi @brosvaby can you please provide a ADB bugreport. We need to understand what the device is running and what the system thinks is happening.

**Comment by wvpaf** [comment]
Hi @brosvaby any luck on the ADB bugreport?

**Comment by brosvaby** [comment]
Actions (please refer to below steps)
1. I downloaded offline license at 12:24
2. I Disconnected Wifi and LTE.
3. I checked duration is 180 sec.
4. I couldn't play content after 12:27
5. However, after I returned the device time to 12:25 minutes, it was possible to play the content.
6. You can always replay the device time back to 12:25 minutes without re-receiving the license
(Attach the report.)
[bugreport-NRD90M-2017-05-15-12-25-50.zip](https://github.com/google/ExoPlayer/files/1000113/bugreport-NRD90M-2017-05-15-12-25-50.zip)


Is this the right scenario?

**Comment by daniel-inka** [comment]
To support DTR(download to rent) scenario, there must be a way to prevent device-time manipulation as above. Maybe it is Widevine CDM's role, or it may be implemented in application level.
Anybody to confirm or advise on this issue?

**Comment by jefftinker** [comment]
Thank you for reporting this issue, we have opened an internal bug against the Widevine CDM and will investigate it further.

**Comment by uthonangi** [comment]
@jefftinker I am sorry but this is stunning. We can reproduce this issue as well and we are in a pretty bad spot. Rentals are essentially broken on WV and it is not even documented anywhere as a known issue on the WV side. Can you please give us an update when you get a minute on the status of the WV bug? I know this forum is not for WV bugs but such a fundamental piece of functionality being broken is disappointing.

**Comment by ram992** [comment]
Hello @ojw28,
Any update on this issue, cause this is haunting my Downloaded content being valid for a particular time.

**Comment by ram992** [comment]
Hey I got one more issue, If the device time is set in future and the license is requested, and if the device time is set to the current one, the license is valid till the end of future time+ it's expiry time period.
Like if for example my license is valid for 1 week, and I fast-forward the device time to one week, then downloaded the license whose validity is one week ,then the validity of the license is 2 weeks if I reset the device time.
How can I resolve this.

**Comment by ojw28** [comment]
This is not an ExoPlayer issue. You need to get a response from Widevine (@wvpaf, @jefftinker, or by communicating directly with Widevine if you have a relationship with them).

**Comment by ram992** [comment]
Hello @ojw28 how do we communicate with Widevine, any info would be great.
And does this communication needs to done at the time of license fetching ??

**Comment by wvpaf** [comment]
Hi if you are a Widevine partner you should have access to our CWIP/partner site. If not this is the next best method of communication.
Investigation continues.

**Comment by wvpaf** [comment]
quick update there is ongoing work to investigate and address this issue. Thank you for you patience.

**Comment by wvpaf** [comment]
potential fix in test now. will provide update on availability soon.

**Comment by ram992** [comment]
@wvpaf  can you comment on the time frame of the availability of the fix, and is it compatible with exoPlayer v2.2.0, as I am not in a position to update to the latest one now.

**Comment by wvpaf** [comment]
Looks like a fix for this case is intended to land in O MR1

**Comment by adesentenac** [comment]
@wvpaf Does that mean that this behavior will stay the same on Android versions before O?

**Comment by fredgc** [comment]
@adesentenac: yes, the behavior for versions O and before will remain the same.  It is unlikely for devices to receive field updates for older versions that are not security patches.
